### ISSUE TITLE
Linux support [help wanted]

### ISSUE THREAD

**Issue by olilarkin** [iPlug2/iPlug2#105]
Reaper now runs on linux and WDL/Swell has GDK support

It shouldn't be hard to make the stand-alone app run with Linux and to then implement LV2 support.  there are two branches for this with the beginnings of the work,  but I could do with help from someone who uses Linux more regularly.

IGraphicsLinux  is the main thing that needs implementing,  possibly using SWELL GDK support

see project: https://github.com/iPlug2/iPlug2/projects

**Comment by olilarkin** [comment]
moving to project

**Comment by AZSlow3** [comment]
I am not sure where to put a comment on that...

Using linux branch I was able to make IPlugMinimal ProcessBlock looping and primary dialog showing (without menu), so APP mode. 

I know, that is not much. But the amount of modifications to make it work was not zero ;)

I just want to say I have plans to make at least app work as desired.

**Comment by olilarkin** [comment]
awesome. after messing around with the makefile, i thought maybe using cmake for linux would be better. on this branch there is some work on that https://github.com/iPlug2/iPlug2/tree/project/cmake

**Comment by AZSlow3** [comment]
cmake can be nice looking and cross-platform solution... when it is already known how everything should be build. I sometimes prefer strait Makefile changes instead of "what is the best way to achieve that with cmake...". Also cmake is not good to compile and cross-compile in one go (my own projects are Linux/Window and I always compile them from one Makefile to be sure nothing is broken after any change).

I will start with Makefiles till app/vst3 works under Linux (with GUI), then we will see how it goes.
For VST2 I have no signed license. I think the best way is to use clean written VST2 ABI header, so not using anything from SDK, including characters "vst" nor other trademarks. Since there is no way to officially use VST2 ABI, as long as licensed files are not used, Steinberg can not enforce anything. They are in Germany/EU, where interoperability precedents exists (attempts to forbid using something, even still officially available, for the purpose of interoperability have failed in curt).

**Comment by olilarkin** [comment]
I don't think VST2 support on linux is important, since it is effectively a dead format. VST3 is actually supported by steinberg on linux, and I think would be a good option for getting iPlug plugins running on ELK OS https://www.mindmusiclabs.com/. APP & VST3 would be great. I would also like to support LV2 eventually.

**Comment by AZSlow3** [comment]
My current region of interest is REAPER under Linux/Windows. So f.e. LV2 will have no value for me (till REAPER start supporting it). I am also not sure I will participate in ELK target.

My another region of interest is Accessibility of iPlug2 (under Windows only, I do not have Apple and Linux is not used by blind musicians). But I have not start looking there yet, while learning how all that staff is working in general will probably help be with that in future.

To stay on topic: I have got IPlugEffect and IPlugControls running with IGRAPHICS_LICE (app mode only). Not everything is working and what I have written is not checked. But app can be compiled and run. GUI appears, updates and can be controlled by mouse. I have not check VST3 and other GUI modes yet.

**Comment by AZSlow3** [comment]
![iPlug2Linux](https://user-images.githubusercontent.com/15633421/67221770-526b1a80-f42c-11e9-85a8-eb691ccf160d.png)

**Comment by olilarkin** [comment]
that is awesome. great work. 

I am also interested in accessibility, and would love to provide a good way for developers to make their iPlug plugin UIs easily accessible to screen readers. Since iPlug can be used with whatever GUI framework on top, one way is to write a GUI using win32 dialogs and use SWELL for macOS/linux version, which seems to have some degree of accessibility support. What would be better is if there could be some automatic way of making an IGraphics UI accessible.

**Comment by AlexHarker** [comment]
Great stuff! If you have any questions about the IGraphics methods please ask here or on the slack channel!

**Comment by AZSlow3** [comment]
At least on Windows, accessibility has 2 scenarios. 

   One and mostly used is widget oriented. Effectively, each "window" with associated text (standard controls have text, f.e. text on button, text in row of list control, title of the window, etc. all in the same property Get/SetWindowText) is already half accessible. Reasonable order of controls and working Tab key to switch between them is sufficient. Heh... some forget about that. F.e. parameters in REAPER generic UI are almost accessible. But Tab is not working, so that view is not accessible. 
   Screen Readers are advanced in interpreting the tree of windows and associated text, that is "primary mode" for them.
   Unfortunately, modern UIs are "single window" from OS perspective. So that method is "no go" for frameworks like iPlug2.

  It is possible to construct a special tree of information about controls on the window, with associated text information. Than it is required to support Tab key to switch and let say arrows to change value, informing which control is in focus. That is called "document mode", like a web page. iZotope provides that (they have not labeled each control and that can be confusing, but still usable)

  Sure, the best way is to add extra information, specifically for accessibility. Like real value text for a fader, even in case the value is not displayed in text. REAPER own plug-ins have that information (and Justin fix related bugs quickly).

  The structure of controls in iPlug2 interface is known by the framework, so making it available to the accessibility system in the OS, with tab/arrow keys operations, should do the trick. But the evil is in details ;)

**Comment by AZSlow3** [comment]
This one took some time... It does not react on mouse yet and crash when closed, but it is animating. There will be challenge to correctly wrap all that into VST3 GUI system since unlike Windows/Apple there is no "common graphic" on Linux.

![iPlug2LinuxNanoVG](https://user-images.githubusercontent.com/15633421/67413468-7dd33e00-f5c1-11e9-97c2-65c6d0d0d883.png)

**Comment by olilarkin** [comment]
bravo! would love to try your work if you would like to push it to an iPlug2 fork

**Comment by AZSlow3** [comment]
I will make NanoVG also react on mouse and close correctly. I will upload the fork then (without VST3)

**Comment by AlexHarker** [comment]
This is fantastic - it looks like you've got font loading working - is that the case?

**Comment by AZSlow3** [comment]
Font loading is incomplete, only named fonts are loaded. But that part is not a big problem with SWELL (or fontconfig directly). Graphic parts are going to be tricky, f.e. "parameter editor" (not implemented yet), VST integration, etc. I am still not sure there will be no show stoppers here, under Linux every "plug-in" makes own "message loop" for everything (over possibly networked connection with graphics, keyboard, etc). So "modal dialogs", like file browser, starts internal loop till it is finished. Effectively in case main program try to do all graphics in one thread, that can block re-draw of any other part of the program. In Windows that does not happened, even local message loop internally process all messages, including other windows (with light consequences only, f.e. dialogs in Cakewalk stop Control Surface updates). Not sure how that works on Mac.

**Comment by AZSlow3** [comment]
I have uploaded current progress to fork: https://github.com/AZSlow3/iPlug2

If someone want check it:
* checkout linux branch
* in dependencies, compile RTAudio and RTMidi in place (./autogen.sh && make)
* change to Examples/IPlugControls and do make (makefile can be used for any example, but I have not checked other)

I have Ubuntu 18.04 LTS. You will need many development packets (gtk3, X, GL, font-config, alsa, pulse, jack, etc.)

What is working so far:
APP only (no plug-ins)
LICE or NANOGL/GL2 graphics (modify Makefile to select)
ALSA/Pulse tested (JACK not)

Explicitly not working yet: dialogs (file/color selection, the code is in place but not launching yet...), parameter editing, tooltips.

Note that you need X11, it will not work on Wayland in GL mode (I do not and probably will never care about Wayland, it is absolutely useless for me...).

Work in progress. The next step is VST3...

**Comment by olilarkin** [comment]
great. i managed to compile it. with the makefile how can we get all the object files etc in a subfolder build-linux?

**Comment by AZSlow3** [comment]
I need compile it for Windows to see correct place for object files, then I can adopt Makefile to follow the same convention.

**Comment by olilarkin** [comment]
intermediates etc go in a sub folder of the project dir e.g. IPlugControls/build-win , build-mac or build-ios on those platforms

**Comment by AZSlow3** [comment]
Just to inform about the progress....

As first, I have tried to understand how to use SWELL in plug-ins. Not easy, it is not designed to work in embed mode. Along the way I had annoying GTK warning. So annoying that I have digged for the reason. That happened to be SWELL problem on Linux, proposed patch is already in WDL master. So that investigation had a positive side effect for REAPER community ;)

Finally I have decided it is better to not use SWELL for plug-in part (for APP host it is ok). May be one day I (or Justin...) make it plug-in friendly, but iPlug pretend to be a framework with platform dependent code. It will not benefit much from high level intermediate multi-platform library.

The next decision was which library to take. SWELL is based on GDK, still too heavy gun just to create GL window for my taste. Xlib is better, but a bit bloated. So, why not xcb? VST-GUI is using it and that is lowest existing level on X. Another plus points are: a possibility to link statically (closed source plug-ins can benefit from better portability, especially taking different Linux distributions and versions into account); reasonable license (GDK is LGPL); less bugs in the library (xcb just wraps X protocol, most code is auto-generated from XML description of X protocol).

But that is not going to be easy. For GL part there is no documentation in the source, no tutorials and almost no examples in other projects. But since it is direct GLX, XLib GL examples should help. I am at that stage now, trying re-write iPlug2 Linux code to work with XCB/GL. That will take a while.

**Comment by jpcima** [comment]
>So, why not xcb? VST-GUI is using it and that is lowest existing level on X.

Hi, from moderately recent experience of Xcb use, based on the LV2 port of Surge synthesizer (VSTGUI), there has been an inconvenience in binary compatibility.

Some Linux variants are shipping different so-numbers of xcb libraries. As consequence, you can pick a build of Surge synth and run it on Debian buster, but can't do so on Debian stretch.

Xlib has been established since a much longer period, it produces more widely compatible binary, which is certainly a benefit for non-opensource plugin makers.

**Comment by olilarkin** [comment]
Thanks for the info @AZSlow3. Are you aware of DPF? another plugin framework that uses nanovg, might have some clues.

Do you have VST3 working without GUI? That would be cool for the ELK. I have one sitting next to me but didn't have much time to play with it yet.

**Comment by jpcima** [comment]
@olilarkin the Linux audio world usually uses a platform window abstraction for embedding and creating the graphics context. [pugl](https://github.com/drobilla/pugl)

In DPF, there exists a highly-modified fork based on a much older pugl version.
Since this time, original pugl author (drobilla) has reengineered pugl in a more capable library; so I think it's better to examine this one instead of what's in DPF at the moment.

**Comment by AZSlow3** [comment]
@jpcima I think good way to avoid compatibility problems is static linking with everything except may be libc (it is more or less compatible between distributions). 

pugl is XLib based. I like to learn things and XCB is a good opportunity for me to learn "original" X and extensions (including GLX). Copy something could be faster, but less fun ;)

@olilarkin I think VST3 by itself (without gui) was working, at least I could load it in REAPER

**Comment by jpcima** [comment]
>static linking with everything except may be libc

It will not be possible, there will be a minimum a libraries which you have to link dynamically.
Just keep it to the smallest possible set.

For instance, libGL comes in different GPU specific flavors, it requires dynamic link.
And then GL dynamic links X11, hence it needs accept dynamic X11 as well.

In DPF, only libGL and libX11 are linked, in addition to C and C++ runtimes.

**Comment by AZSlow3** [comment]
It is clear GL/GLX is not going to work without XLib. 

But there is a "trick". Libraries can be run-time loaded instead of dynamically linked. Then program/plug-in can be loaded even when the library in question does not exist or problematic. I think for audio plug-ins it make sense. They can work without GUI. So, why make them strictly dependent from some GUI library? Many plug-in types can be used in "headless" mode not worse then with GUI, driven by exposed parameters. Also someone can come to idea render projects on remote server/cloud...

Is that technically feasible? I think yes. iPlug2 already make use of glad for GL. glad can also generate run-time loader for GLX. So (while not yet tested) that part is covered by existing tools. 

Several XLib functions required to make initial connection can be loaded manually. That I have already tested, I have a XCB (statically linked) program which initialize XLib (run-time loaded) connection. XCB (+ 2 small libraries from X world), for this small test with just one line window is 200k. I remember statically linked XLib even long time ago was way more. Also since it is going to be loaded by GLX it make no big sense. 

For the moment there is just one problem for which I do not see any solution... XLib/XCB can only work in case XLib itself is XCB based. That is true for Ubuntu, but I am not sure that is true for all existing distributions...

Finally, it will be great in case plug-in can fallback from GL to raster based back-end. I remember on Windows there was such plug-ins and that was helpful (in case GL is problematic, plug-in is made for too recent GL, etc.). In case iPlug2 framework allow that, XCB can fallback to XCB pure connection and continue to work with GUI completely independent from the X environment (including remote X connection, where GLX is tricky for end user).

**Comment by AZSlow3** [comment]
I have proved the concept... While I had to modify generated glad_glx, I have managed to compile 170kB test (not iPlug2 yet) which is linked with libc and libdl only. It shows working GL child window when everything is ok, but it also run when there is no graphic libraries (obviously without any graphics, but that means audio processor will be able to work headless).
The next step is put that into iPlug ;)

**Comment by AZSlow3** [comment]
Update:
![iPlug2LinuxVST3](https://user-images.githubusercontent.com/15633421/69883676-7fd1a080-12d5-11ea-8119-d974c18c8505.png)

I just had to make it show up... for myself, to prove I can manage that. After all that digging I was a bit frustrated. But there is still one big "hack", this VST3 gui can show up in GLib based hosts only (f.e. REAPER) ;)

I will have to touch iPlug2 core part now. The problem is Linux specific. There is no "implicit main loop". So, it is in general not possible to create "a timer" or "a window". Currently iPlug2 assumes that is possible. Note that SWELL is not prepared to deal with that problem, it has its own GDK (so GLib) based loop, in addition SWELL arbitrary call it when opening dialogs. That theoretically can be started in a dedicated thread, but I do not like the idea. That is not what hosts expect from plug-ins.

There are several scenarios:
* standalone APP. SWELL is written for that and so it can handle APP part. Plug-in still embed X Window as such, to avoid differences between incarnations. That can work with direct "hack", I am embedding plug-in internal event processing directly into GLib. I could use polling using SWELL timer instead, SWELL itself is not using infinite GLib main loop. But I had to implement "right" embedding anyway (see VST3 later) and integrating with GLib is just several lines (and does not required linking with glib).

* VST3. Steinberg has recognized the problem. So they have introduced extra interface for Linux, exactly to integrate timers and graphics events into host main loop. VSTGUI is using that method and I think that is the only good solution. I will modify iPlug2 classes to support that. Apart from plug-in GUI, iPlug2 has "Timer" class which should work throw the same hook.

* LV2 (and I guess VST2): while LV2 is "Linux native", I have not found standard extension for main loop integration. But there is polling extension, I guess GUI and timers will have to use that.

I am going to implement VST3 main loop integration and try to finish all mouse operations (since I use XCB directly, that is not as trivial as with SWELL) before uploading.

So, my current approach for plug-in GUI is XCB (possibly statically linked, proved to work) + XLib/GLX (run-time linked, proved to work) and SWELL for APP (main window). I will stay GL only for a while. XCB+Cairo should be possible. LICE without SWELL is not checked yet.

Strategically I think making dual-GUI plug-ins is a good adea. I mean plug-in which can run in GL, but can fall-back into other mode (like LICE) when GL initialization fail (or the user wish it does not use GPU 3D, the subject of different power consumption and so heat and noise, especially on notebooks). Current iPlug2 framework is not supporting that (only one compile time IGraphics impementation).

**Comment by olilarkin** [comment]
thanks for your continued work on this!

**Comment by AZSlow3** [comment]
Uh... Very Stupid Toolkit level 3 (VST3) does not make the live easy for users. No timers, no onIdle. So no way to make IPlugAPIBase::OnTimer called without GUI open. Really, NO WAY (till someone can prove me wrong). And from answers on St forum, that is by design.

In other words, processor and controller in that framework can not do anything on there own wish.

I think the paradigm is "simple": the processor has to process blocks "in vacuum", with all relevant information delivered as parameters and all changes exposed as parameters (so, IPlugAPIBase::OnTimer activity supposed to be inside RT processing). That IMHO creates many "chicken and egg" like problems, including particular host handling (driven by almost not existing documentation and rather bad SDK helper classes implementation). For some there are workarounds (of different hackish quality), some was "no go" till (after years) St has recognized that and implemented addition (f.e. all MIDI messages "Legacy" output), some still have no good solutions (or it is difficult to find). 

The explanation.
Windows/Apple have "default" main loop. So creating a timer or GUI elements is not a problem, they will be correctly attached to corresponding thread(s) main loop(s) "for free".
VST2 has onIdle() call from the host, in correct thread. So related activity could be easily emulated (with some performance overhead) on systems without well defined main loop. 

Once St has decided to support Linux, they hit the problem of main loop (timers and GUI events). So they have introduced  "Linux specific" interface, IRunLoop. It could be nice to have it for the host, so not GUI related. That could help create timers and async IO in more or less standard way.
But Very Stupid Toolkit does not like easy ways. They have added it to the GUI only, as IPlugFrame extra interface. 

How that work in SDK? It rather interesting sequence:
1) after IPlugView is created, its setFrame method is called, with host defined IPlugFrame. The rest happens in vstgui (not core part)
2) this IPlugFrame is wrapped into FrameConfig
3) the frame is opened with this FrameConfig
4) RunLoop is initialized in the Frame ctor

After that, timer API can be used in GUI (throw global RunLoop).

So they know all that is ugly, they could not find more or less elegant way to use own "creature"!

----------------------
I give up  IPlugAPIBase::OnTimer for VST3. It can be implemented for GUI, theoretically it can be host dependently hacked (I already have gtk hack, i guess there is similar way for Qt...). But general correct implementation is not possible on Linux.

**Comment by AZSlow3** [comment]
May be you can give me a hint for correct modification for GUI...

IPlugVST3_View overrides setFrame, where it gets IPlugFrame (VST3+Linux specific).  It can be converted to my (VST3+Linux specific) structure and should call IGraphicsLinux method (Linux, but not VST3 specific) with this structure. So I need pass it throw IGraphicsEditorDelegate, probably some "neutral" way. Probably the same for finalization (when the View is asked to forget about Frame).

Doing all that directly looks as ugly as SDK behavior explained in the previous post ;)

**Comment by olilarkin** [comment]
could you update your branch so i can see where you are at with the timer stuff?
IPlugAPIBase::OnTimer is a main thread timer, Not sure it's relevant but here is vstgui's linux timer
https://github.com/steinbergmedia/vstgui/blob/develop/vstgui/lib/platform/linux/x11timer.cpp

**Comment by AZSlow3** [comment]
Once I put it somewhere (I mean main loop integration), I will upload.
My long explanation before is about "a main thread timer". It is not supported in VST3 (it works on other platforms just because they create objects in main thread, OS provides timer and hosts run main loop on this thread. Nothing from that is really documented as a must.
I have also explained that vstgui's linux timer work when the GUI is open and the plug-in is VSTGUI4 based.  In other words, it is not main thread timer and it can not be used in other then VSTGUI frameworks. It is not "Interfaces -> SDK helpers -> [VSTGUI]", it is "Interfaces -> VSTGUI -> SDK helpers". That is why I call that "ugly" (apart form several different "RunLoop" classes/structures and other equally named classes in X11:: and Linux:: namespaces, including X11:: to Linux:: "mapper", which make it very hard for a user to understand global logic).

**Comment by olilarkin** [comment]
Apologies I’ve been a bit busy and distracted with other work sorry I didn’t read your messages properly

**Comment by AZSlow3** [comment]
"When you go deeper into the forest, you meet more trees..."

After somehow integrating the processing, I have noticed that REAPER is not calling registered timer. Also it never really check registered connection is ready to read, it just calls "its ready" every 20ms. So the implementation is still onIdle(), but wrapped into VST3 interface ;)

Since I was not sure the problem is not on my side, I have decided to check with SDK editorhost. Ha! LOL! OMG... https://github.com/steinbergmedia/vst3_public_sdk/issues/16
Not sure how long my post is going to survive, so in short: editorhost for Linux was written by someone who had no time/was drunk/was never programming event loops for linux before. I mean it works, consuming 100% CPU and not really checking what it should (they do not watch the connection in question nor set timer correctly, but busy looping and so triggering plug-in).

Well, as the result. With fixed editorhost, my approach works  correctly there (with events and timers). It also works with REAPER (fortunately what it does is ok for the way I process timers and events, I mean when timers are not triggered explicitly). I am going to cleanup the code and upload soon (without finishing other staff).

**Comment by AZSlow3** [comment]
Sorry, no upload this week... Just before committing I have decided to compile Controls example. Apart from crash in case of selecting in pop-up menus (with that I still could upload and fix later), bitmap rendering does not work. SVG and the rest works (as APP and VST3), but no bitmaps (not even in APP, where it was working with initial incarnation). That is serious. I have connected NanoVG directly to my test app, without iPlug2, and bitmaps also do not work there. GL produce some low level warning in debug mode, but it also produce it when I do not try to use bitmap patterns. Google keep silence about that warning.
After several hours I start to think that can be the consequence of my glad_glx trick, but I do not have time to check next days.

**Comment by AZSlow3** [comment]
I have uploaded the progress. More or less works.

The issue with images comes from re-entrant GL context activation in SetScale. In Windows code it set "current" and then restore "saved", which in this case just set context to what it already was (and restore it to the same...). My code was a bit different, so it was resetting the context before created bitmaps was saved into GPU.
I no longer reset the context. For some reason, all X events (f.e. mouse) collected during drawing are discarded in case the context is reset before these events are processed (I guess a bug somewhere in underlying libraries).

It is still periodically crashing in "popup" menus. iPlug2 does not check the menu is there on MouseMove, believing corresponding state is correct. It seems like that is not always the case, so it is calling zero object with logical result (crash).

Also it seems like something is wrong with "updates" logic. If I start to update every 2 seconds instead of 20ms, some part is lagging (and not updating at all till the next change).

I have added Dirty to knobs when mouse is clicked. They is visual difference when the button is clicked (activated), but since it was not declaring Dirty on click (only on release) it was not shown.

APP incarnation should work as usual (SWELL host with timer). VST3 has no "timer", so "Idle" is not called. I could do this when GUI is open (so timer is available), but that make no big sense in long term.

**Comment by olilarkin** [comment]
great! thanks for this i will check it out

**Comment by olilarkin** [comment]
i pulled your latest changes but seem to be missing IPlugVST3_RunLoop.h/cpp

**Comment by AZSlow3** [comment]
Sorry, added

**Comment by AZSlow3** [comment]
No functionality changes, but I have completely reworked Makefiles.
So after cloning github, in Examples/IPlugControls or Examples/IPlugEffect, it is possible to type:
`make -f ../Makefile VST3`
(or APP instead of VST3, and/or add -DIDEBUG=YES for more messages in console)
It should download/compile required dependencies automatically.

VST3 is created as a "bundle folder", APP works with external libSwell.so (it is compiled and put into correct place).

This Makefile is dependency aware, so it works correctly in case IPlug2 sources are modifies, options changed, etc. It is also prepared for complicated compilation (different flags for different files), several goals for one target (f.e. for LV2). It was thought to be "one for all", that is why I put in in Examples instead of particular plug-ins. Also "external plug-ins" can be compiled with make -f <path to this Makefile>, without any other changes.
The results are saved into Dependencies/Build (not touched once created) and Examples/Build (rebuild plug-in incarnation from scratch when needed).

I want check LV2 also can somehow works in that schema, before I continue with graphics.

It seems like whatever effort VST3 had to force "own rules", stable solution should have a separate own thread for "main" and "GUI" tasks. Per plug-in instance (can be resource wasting and real-time performance disturbing with many plug-ins in a project), per plug-in (as VSTGUI does with X binding) or per framework (so one global iPlug2 main/GUI thread). Probably the approach should be configurable. Problematic part is thread safety (between "real" main thread and this one, between RT thread and this one and between different instances when in shared mode).

**Comment by olilarkin** [comment]
thanks again... hope to look at this properly soon

**Comment by AZSlow3** [comment]
It is not yet worse to spend your time on that... But I am slowly moving forward. 

Already auto-generating LV2 ttl files for IPlugEffect, so Ardour could find it,  load it (as LV2) and build default GUI. But processing is not yet implemented at all.

PS. LV2 wants parameters description in static text files, but parameters are defined inside user code. Duplicating this in a separate description is not nice. So I have checked that I can compile "config" executable first, which generates that files after instantiating the processor.
I have not really checked your LV2 branch, it can happened I am re-inventing the wheel...

**Comment by olilarkin** [comment]
i didn't get far but planned on borrowing DPFs TTL stuff

https://github.com/DISTRHO/DPF/tree/master/utils/lv2-ttl-generator
https://github.com/DISTRHO/DPF/blob/master/utils/generate-ttl.sh

**Comment by AZSlow3** [comment]
Yes another incarnation of iPlugEffect (LV2 in Ardour):
![iPlug2LinuxLV2](https://user-images.githubusercontent.com/15633421/72628110-f9892080-394d-11ea-80f0-9aa27acf50a3.png)
That is not complete LV2, just bare minimum for iPlugEffect (but functional, the plug-in change the signal, react on knob turning on own interface and host auto-build interface).

I plan advance XCBT and probably will test it with iPlug2. Probably I will also check what can be done for accessibility. 

But I do not plan participate iPlug2 development in general. Sorry to say, but iPlug2 by itself is too nontransparent with strong VST2 oriented "flavor" for my taste. Especially on Linux that is too dirty. In particular:
* VST3/LV2 strictly separate processor from UI. Any common (or looking like common) data and methods is looking to troubles. So PLUG_CLASS_NAME should not include both. 
* the concept of "general timer" is error prone, here I agree with Steinberg. There are three distinct activities in plug-ins: audio processing with current sample driven "time", background activity without any timer needs (f.e. samples loading), GUI activity which should follow real time and sample time. Mixing anything from these three is again looking for troubles
* I think single binary plug-in should support several graphical back-ends when possible. The same for different features in different hosts. Up to "all in one" binary for several plug-in formats (I am slowly developing one plug-in which already support VST2 and VST3 ABI in one binary, now I plan to add LV2 to the list).
* special care about multi-threading means clear definition what can work where and when. VST3 is already too fuzzy in that respect (f.e. REAPER loads plug-in state after activating real-time processing), iPlug2 adds extra complexity with overloaded classes. I think on Linux the only reasonable general concept is allow background and GUI work in separate threads. That means DAW GUI+Main thread(s) and plug-in GUI+Background threads are not the same (can run in parallel).

**Comment by olilarkin** [comment]
awesome work. Thankyou for this. I understand your reservations. We have a solution to build distributed VST3s, but it is not a very nice solution, involving compiling a static lib for the controller part with PLUG_CLASS_NAME modified. This is a problem of trying to keep a very minimal plug-in implementation and also support complex plugin APIs. If I were a linux guy, I would just make LV2s

**Comment by olilarkin** [comment]
everything builds 👍  will try later - that is one hell of a makefile!

**Comment by olilarkin** [comment]
I got audio working with VST3, but so far no GUI on my ubuntu VM

**Comment by olilarkin** [comment]
BTW skia is now our preferred graphics backend and it can do either software or gpu rendering. At the moment it’s a compile time switch but we could make it runtime switchable

**Comment by AZSlow3** [comment]
> I got audio working with VST3, but so far no GUI on my ubuntu VM

Is glxgears working?

I have just re-build using a clean clone and it works for me. Have you checked APP? Does VST3 produce some warnings during compilation (except several "DEPRECATED" from VST3, I do not see any warnings)?

Can you compile (in Examples/IPlugEffect) with:
make -f ../Makefile IDEBUG=yes VST3
and post what you see in the standard output when you try to open the GUI.

My approach is loading GL dynamic (so no linking deps on GL), the plug-in will not fail in case that is not possible but GUI will not work since there is no "fall back".
 
I will look at skia.

**Comment by olilarkin** [comment]
I tried glxgears and it works

The APP just shows grey GTK window and produces the following std output:

```
oli@ubuntu:~/Dev/iPlug2/Examples/IPlugEffect$ ../Build/Linux-x86_64/IPlugEffect/APP/IPlugEffect 
BEGIN IPLUG CHANNEL IO PARSER --------------------------------------------------
Channel I/O #1 - 2-2
               - input bus count: 1, output bus count 1
               - channel count on input bus 1: 2
               - channel count on output bus 1: 2
               - input channel count across all buses: 2, output channel count across all buses 2

1 I/O configs detected
Total # in chans: 2, Total # out chans: 2 

END IPLUG CHANNEL IO PARSER --------------------------------------------------
Reading ini file from /home/oli/.config/IPlugEffect/settings.ini

RtAudio Version 5.1.0
APP: Initial socket size: 600x600
IGraphics NanoVG @ 60 FPS
couldn't attach gui
```

The VST3 just produces an empty black window with editorhost and 
```
oli@ubuntu:~/Dev/vst3sdk/build/bin/Debug$ ./editorhost '/home/oli/Dev/iPlug2/Examples/Build/Linux-x86_64/IPlugEffect/VST3/IPlugEffect.vst3' 
BEGIN IPLUG CHANNEL IO PARSER --------------------------------------------------
Channel I/O #1 - 2-2
               - input bus count: 1, output bus count 1
               - channel count on input bus 1: 2
               - channel count on output bus 1: 2
               - input channel count across all buses: 2, output channel count across all buses 2

1 I/O configs detected
Total # in chans: 2, Total # out chans: 2 

END IPLUG CHANNEL IO PARSER --------------------------------------------------
IGraphics NanoVG @ 60 FPS
```

**Comment by olilarkin** [comment]
is it possible to build with LICE at the moment?

**Comment by AZSlow3** [comment]
Sorry, the command should be:
make -f ../Makefile IDEBUG=YES APP
so "YES" in capital to produce more debug output. It should print at least something useful when failing...

LICE is not supported so far, it is going to be the next "step" (in case that is possible without SWELL).
SKIA / GL2 "somehow works". I have not uploaded that change yet, since I wonder how it works... Related code is compiling (I had to add skia includes, library, GL, freetype. Also I had to fix one source since some skia constant names was changed. And I have explicitly added a bug in NanVG, to check it is not participating the build...) and the picture is not the same as with NanoGL. Still, I have not expected it working without any Linux specific mods.

**Comment by olilarkin** [comment]
ahhh...

```
oli@ubuntu:~/Dev/iPlug2/Examples/IPlugEffect$ ../Build/Linux-x86_64-dbg/IPlugEffect/APP/IPlugEffect 
BEGIN IPLUG CHANNEL IO PARSER --------------------------------------------------
Channel I/O #1 - 2-2
               - input bus count: 1, output bus count 1
               - channel count on input bus 1: 2
               - channel count on output bus 1: 2
               - input channel count across all buses: 2, output channel count across all buses 2

1 I/O configs detected
Total # in chans: 2, Total # out chans: 2 

END IPLUG CHANNEL IO PARSER --------------------------------------------------
Reading ini file from /home/oli/.config/IPlugEffect/settings.ini

RtAudio Version 5.1.0
APP: Initial socket size: 600x600
IGraphics NanoVG @ 60 FPS
INFO: compatible GLib is not found, attempt to load GLib
INFO: GLib is loaded explicitly
Loading atoms
 [0] WM_PROTOCOLS: 0x123
 [1] WM_DELETE_WINDOW: 0x121
 [2] _XEMBED_INFO: 0x15d
Xlib/XCB FD: 9
Found 12 usable FB configs, choosing best for MSAA
BUG: sutable FB Config(s) found, but could not be used
BUG: attempt to unregister unknown window
BUG: bad parameters for window_set_handler
Not exist
couldn't attach gui
```

**Comment by olilarkin** [comment]
I have merged your work on the linux branch and made a couple of changes

https://github.com/iPlug2/iPlug2/commit/6a0807ac138d6cf04010700d9d461572957b6e9f

https://github.com/iPlug2/iPlug2/commit/65fc39479a475dd48b9327a5d4c27d943bbea464

**Comment by AZSlow3** [comment]
Well that was too early... I (hopefully) has fixed not working graphics. I have installed vanilla Ubuntu 19.10 in VM and could reproduce. There was several bugs (typo, potential GLX version dependent problem and system library naming). Really the effect was from one character typo, but it is good I have looked around ;)

IPlugProcessor exclusion is nice, thanks for good solution!
Timer protection removing is fine now, when I know where and when it is used.

Merging with upstream was not without consequences: I had to change options and recompile SKIA.
But not yet solved is magical segfault in RtMIDI initialization. I have just commented corresponding line for now. The diagnostic is really "magical": in MidiIn method it complains and crash in MidiOut type check. Can be some GCC interpretation of RtMIDI C++ wrappers bug. I have not looked deep.   

At least APP (but VST3/LV2 as well, I just have not checked after all merges) should also work with IGRAPHICS=SKIA_GL2. And the result looks better then NANOVG_GL2. Unfortunately that adds compile time deps to GL, expat and freetype. GL I do not really like, but I am not sure SKIA can be convinced to work throw GLAD.
Note that SKIA compilation takes quite some resources (and time). As with the rest, it should be auto-downloaded-compiled (once and only when asked by IGRAPHICS). It can crash on VM (out of RAM if VM has many cores), I have foreseen SKIA_NINJA_ARGS. So f.e. "-j 2" for it theoretically can help (not tested, I had to use -j 4 at the end on my 8GB RAM i7 when compiling complete debug version, used in iPlug2 flags do not compile binaries nor debug, so probably all that is void for iPlug2).

**Comment by olilarkin** [comment]
great, now the UI works in app and vst3 :-) i didn't try lv2 yet. SKIA is indeed painful to build, but it is an excellent library. On other platforms we provided pre-built static libraries, but I guess for linux it's best to build locally

**Comment by olilarkin** [comment]
skia works also

**Comment by AZSlow3** [comment]
IGRAPHICS=LICE is also possible now.
It is "prototype quality" implementation, I do not even check for bitmap format compatibility. But it works for me.
Note that result are visibly different for APP and VST3/LV2. In APP LICE is using normal SWELL with GDK based font rendering. In VST3/LV2, LICE is using headless (compiled in) SWELL with freetype font rendering.
In both cases LICE is rendering into memory bitmap, which is transferred by X (so slowest possible method). XShm is faster and easy to add, if LICE is going to be used more then just for testing.

**Comment by AZSlow3** [comment]
CAIRO.
iPlugControls flickering (not really visible with iPlugEffect).

But there is something wrong with Cairo code. I was getting bad crashes and I could trace them to leaking reference counter for mSurface and its device. XCB tend to return the same address when re-connecting fast (f.e. on plug-in change in FX browser of REAPER), Cairo is caching XCB connections with bizarre results (sometimes crashing in assert, sometimes generating bad X requests and drawing nothing).

I use cairo_device_finish() to force kill the reference. I do not think that is Linux specific, but you can check (see commented printfs there for possible reference counters, I get 3 for surface and 11 for device, so even mSurface is not really destroyed by "destroy" call).

**Comment by olilarkin** [comment]
thanks, i will check it out soon ;-)

**Comment by AZSlow3** [comment]
I have switch to double buffering for CAIRO. IPlugControls looks reasonable now (checked in APP and VST3 incarnation).

**Comment by AlexHarker** [comment]
You should note that with Cairo if there is a context referring to a surface then calling destroy on the surface then It will just decrement the reference count and it won't get destroyed till later.

**Comment by AZSlow3** [comment]
That is why I write "something is wrong with Cairo code". There can be some Linux specific features, f.e. when XCB surface is destroyed right after creation, its Device reference is stay 1 (probably they think "X connection is the same till the program exit", wrong assumption for on-demand case). But surface reference can be general problem. That is why I have suggested to test on other platforms.
I mean I do not understand your point...

For the topic. AGG is also supported now (prototype style, as LICE, in practice it works slower then LICE since it does "anti-grain").

In general, only 24bit per pixel / TrueColor with LittleEndian image byte order works with LICE and AGG at the moment. In practice, I guess till someone tries to use exotic devices for audio production, there are only 2 cases: x86 with local graphics (covered) and Raspberry PI (to be understood).  So I guess it is ok not care about exotic visuals (15/16bit and all not TrueColor) even on alpha/beta/release stage. But I am open for other opinions.

To not leave "holes", I am going to add VST2 as possible target and xx_GL3 graphic targets before declaring prototype complete.

For alpha stage I will need to add at least keyboard support and cursor changes, File selection can be done with VST GUI style workaround (calling external file browser, which normally exist on Linux systems). "Tips" are not critical, the same for platform specific parameter editing and pop-ups.

The biggest question is what to do with timers/"on-idle". VST3 and LV2 do not support that, except for graphics.

**Comment by olilarkin** [comment]
In VST3 plug-ins (currently only with "distributed vst3" IPlugAPIBase::OnTimer() is not essential, because The host is supposed to deal with message passing from processor->controller and controller->processor. So that just leaves OnIdle() and whatever the plug-in developer wants to do there. The most likely thing to do there is to process queues to send data from processor to controller for UI visualization [e.g. here](https://github.com/iPlug2/iPlug2/blob/master/Examples/IPlugInstrument/IPlugInstrument.cpp#L65 ), but that is only necessary when the UI is open.

**Comment by AZSlow3** [comment]
I will think more about Timers...

VST2 (without REAPER extensions,  when reported as supported, REAPER opens plug-in GUI using SWELL pointer... that obviously works for SWELL based Cockos own plug-ins only...).
Note that VST2 header needs well know one line modification to be usable on Linux...

So, current version covers:
* APP/VST2/VST3/LV2 targets (LV2 at bare minimum for simple effects, so no MIDI, no sample accurate parameters, no preset saving)
* SKIA_GL2/3, NANOVG_GL2/3, CAIRO, LICE and AGG graphical back-ends. No resize, strict limitation on X Visuals for LICE and AGG.

I have not tested all combinations, but all of them supposed to work. As I have mentioned before, APP+LICE is a bit different from <other targets>+LICE. All other graphical back-ends work the same way in all targets.

I have learned a lot by this excesize ;)

**Comment by CBiX** [comment]
LV2 support would also open up the [MOD platform](https://moddevices.com/) and ecosystem for iPlug audio devs. The plugins run headless and are remote controlled by an HTML-based UI (and knob/MIDI assigns), which means less trouble porting iPlug's UI part, however it needs a separate UI only for MOD. Maaaaayybeee the HTML-based UI would also work with WebGL or Canvas so you could reuse existing WAM code but it's unlikely...

**Comment by olilarkin** [comment]
hey @AZSlow3 . Do you know how to enable HighDPI mode in linux?

**Comment by AZSlow3** [comment]
From my knowledge Linux does not have system wide HighDPI. Every framework, desktops, etc. (Gnome, KDE, XFCE, GTK2, GTK3, etc.) have own scaling settings. And they are on there own to scale.

**Comment by mxa** [comment]
The title says help wanted, what kind of help is needed?

**Comment by actonDev** [comment]
Not much to contribute in the discussion, but wanted to tell you that I've been working on building with `meson` build system (wanted to have it cross platform, never used `cmake` etc, so `meson` seemed really nice cause of the documentation)

## Windows 
Can build `app` & `vst3` on windows (`IPlugControls` example)

The code for the `iPlugControls` can be seen [here](https://github.com/actonDev/meson-iPlug2) and the changes that were done to the `iPlug2` repo to enable the `meson` build can be seen [here](https://github.com/iPlug2/iPlug2/compare/master...actonDev:feature/meson_build) 

## Linux
Can build `app` on linux (`IPlugControls` example)

However the window is not appearing upon launching the application. In the console I'm seeing the following
```
SWELL: swt 'IPlugControls'
APP: Initial socket size: 1024x768
IGraphics NanoVG @ 60 FPS
couldn't attach gui
 ```

One weird thing that I noted is that I don't get the `SWELL: swt 'IPlugControls'` message when running the application using the `Makefile` of @AZSlow3  (with which I could build & run `IPlugControls` just fine)

----
So if anyone has any idea about the gui not appearing would be appreciated. The work in progress about building in linux is on the `feature/meson-linux` branch in both of the repos mentioned above.

**Comment by AZSlow3** [comment]
Graphics is significantly specific on Linux, my branch use many workarounds to make it somehow work. But nothing is polished nor well tested.

**Comment by actonDev** [comment]
Wanted to give some heads up.
I'm working on a VST3 with a custom graphics backend (I'm using ImGUI) and have it working on windows & linux.
I'm using `meson` build system. I took things from the `linux` branch, the minimum of required changes, cause I also wanted to make the diff quite easy to see.
I skipped plenty of things from the `linux` branch since I wasn't working with any `IGRAPHICS`.

Here's the diff https://github.com/iPlug2/iPlug2/compare/master..actonDev:feature/meson-and-linux-nogui

@olilarkin 
Please check it if you can. Excluding the `meson` related files, the relevant changes (`.h/.cpp`) boil down to 18 files.
Some of them are for the the `LV2` functionality, which I haven't checked yet.

I know that `CMake` is more popular but I hadn't ever used it and I found `meson` a pleasure to work with. I think you'll find
the build definition quite readable.
In case you'd consider adding `meson` support in your repo I'd be happy to help. Till then, I'll be working with my fork. Just merged upstream today and all is still well.

**Comment by olilarkin** [comment]
great to still have people hacking on iplug linux, i hope to look at it soon :-)

**Comment by atilde** [comment]
Does anyone have an up-to-date list of what does and doesn't work or at least a set of tests that can help me quickly identify a good place to start helping out?

**Comment by David-Else** [comment]
Any progress to report on Linux support from anyone? I saw a couple of Linux related commits https://github.com/iPlug2/iPlug2/pull/873 https://github.com/iPlug2/iPlug2/pull/881

**Comment by olilarkin** [comment]
Not really. I will try and get the CMake stuff in first: https://github.com/iPlug2/iPlug2/tree/cmake

Pretty encouraged that [BlueLab were able to use the linux branch to make loads of plugins](https://github.com/deadlab-plugins/bluelab-plugins), however I would like to clean it up a bit before merging

**Comment by David-Else** [comment]
@olilarkin This sounds like great news to me! The reason I ask is that the amazing https://www.fullbucket.de/music/vst.html plugins don't work on Linux, and the website says:

> The plug-ins on this site are freely available for both Windows/macOS and are built using the [WDL-OL/IPlug](https://github.com/olilarkin/wdl-ol) and [iPlug2](https://github.com/iPlug2) frameworks maintained by Oli Larkin and community.

Do you think it will be easy for him to release Linux versions when you have done the cleanup and merge? Any idea of the time frame for that? Thanks for all your great work! :)

**Comment by olilarkin** [comment]
It's not really a priority for me to do that work since I'm not using Linux much. Also I haven't much spare time so I still hope some dedicated Linux person will take up the challenge. Perhaps if there are enough Linux users interested the work could be crowd-funded?

**Comment by David-Else** [comment]
@olilarkin Cheers, thanks for the info.

Crowd funding sounds like an exciting idea! Do you have an approximate amount of money you would need in mind?

**Comment by mxa** [comment]
If I may recommend someone for the job, I know a skilled Linux dev with experience in cmake, VST3 and audio plugins who is available for the job.

**Comment by 38github** [comment]
I will toss in money for Linux (Glibc/musl) support!

**Comment by AlexHarker** [comment]
As there is talk of specific people and or money can you get in touch with @olilarkin and I to see the best way to arrange this - we have some funds now but there might be a way to make a pot for this and have a person/people in mind working on it. I can run in a virtual machine, but I haven't tried any iplug2 stuff there yet...

**Comment by tank-trax** [comment]
I have a build environment on Debian 11, but haven't tried compiling this. I can help with testing

**Comment by olilarkin** [comment]
I need to refresh my memory of it, but IIRC the IPlugInstrument example on the linux branch has a makefile and cmakelists.txt that might be useful. also see Documentation/cmake-doc.md on that branch

https://github.com/iPlug2/iPlug2/tree/linux/Examples/IPlugInstrument

**Comment by 38github** [comment]
Does anyone know how we can find people that would be interested in working on Linux compatibility completion? Has anyone used Fiver to find people to do things like this?

**Comment by recallmenot** [comment]
Hi, I want to join in but I'm new to iPlug. It's fantastic that yall are trying the hard work of bringing this to linux & cmake.
I'm on manjaro, all the package dependencies should be installed.
So far I've cloned the repo, linked in the vst3sdk like this:
```
ln -s ../../../../vst3sdk/base/ base
ln -s ../../../../vst3sdk/pluginterfaces/ pluginterfaces
ln -s ../../../../vst3sdk/public.sdk/ public.sdk
```
then in the IPlugEffect dir, when I run `cmake -S . -B build-linux -DCMAKE_BUILD_TYPE=Debug`, I get
```
-- Configuring done (1.5s)
CMake Error at CMakeLists.txt:50 (add_executable):
  Cannot find source file:

    /home/xy/VST/VST3/iPlug2/Examples/IPlugEffect/resources/main.rc_mac_dlg

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc


-- Generating done (0.0s)
```
What am I doing wrong?
there is a main.rc file in that dir, I guess the main.rc_mac_dlg should be generated from it?

**Comment by olilarkin** [comment]
The macOS/Linux standalone app in iPlug2 uses Cockos' SWELL to emulate the win32 API. The file resources/main.rc needs to be converted for these platforms using swell_resgen.sh (in the WDL/SWELL folder). This should be set up for IPlugInstrument on the Linux branch I think. Maybe it is not set up for IPlugEffect

**Comment by 38github** [comment]
> What is the state of this @olilarkin ? I am willing to help but unsure what where to start with linux support.

Too bad you recieved no reply. I would love to see Linux support.

**Comment by olilarkin** [comment]
The reality is, I have barely any time to work on it, and it's only going to happen if we find the right person who is really knowledgeable about Linux to tidy the existing implementation and maintain it, since I don't have time for that. I'd love to be able to make it support Linux well, But at the moment it's not realistic for me to do that.

**Comment by RustoMCSpit** [comment]
> The reality is, I have barely any time to work on it, and it's only going to happen if we find the right person who is really knowledgeable about Linux to tidy the existing implementation and maintain it, since I don't have time for that. I'd love to be able to make it support Linux well, But at the moment it's not realistic for me to do that.

can we pin the issue or add that we need help to the readme? we can post around reddit and lemmy looking for help

**Comment by RustoMCSpit** [comment]
> What is the state of this @olilarkin ? I am willing to help but unsure what where to start with linux support.

> The reality is, I have barely any time to work on it, and it's only going to happen if we find the right person who is really knowledgeable about Linux to tidy the existing implementation and maintain it, since I don't have time for that. I'd love to be able to make it support Linux well, But at the moment it's not realistic for me to do that.

@diegodorado ???

**Comment by RustoMCSpit** [comment]
oli you see how in zed they have the linux issue pinned up top, can we do the same? https://github.com/zed-industries/zed/issues

**Comment by tkna91** [comment]
@RustoMCSpit You mean this? https://github.com/zed-industries/zed/issues/7015

**Comment by RustoMCSpit** [comment]
yeah you see how it is pinned to the very top, id like to do the same with this issue in this repo. this is one of the most important issues and it needs help and attention according to @olilarkin so doing that will bring attention. @diegodorado has said they want to help so im waiting for them to get in touch with oli

**Comment by olilarkin** [comment]
I have pinned the issue! thanks, i didn't know about that feature of github

**Comment by RustoMCSpit** [comment]
please also add it to the top of the readme that this needs attention

**Comment by olilarkin** [comment]
It's fine as is

**Comment by olilarkin** [comment]
Good news... I have a friend who is going to help me get going again with the Linux support.

**Comment by RustoMCSpit** [comment]
> Good news... I have a friend who is going to help me get going again with the Linux support.

![image](https://github.com/iPlug2/iPlug2/assets/134429563/547c58c5-e577-4a0e-9026-e62c0a045823)

**Comment by RustoMCSpit** [comment]
> Good news... I have a friend who is going to help me get going again with the Linux support.

i know this is so early but what is the expected duration (as in, will it take multiple years or far less)

**Comment by RustoMCSpit** [comment]
> What is the state of this @olilarkin ? I am willing to help but unsure what where to start with linux support.

also oli please get your friend in touch with this person, they also want to help :D

**Comment by olilarkin** [comment]
Absolutely cannot make any promises about when it happens. I already have
no time to work on it. And many other things demanding my attention!

I want to get the CMake Stuff merged, Then get a very basic Linux
implementation merged. Hopefully that is enough for somebody to pick up the
work If they really want this feature.

**Comment by RustoMCSpit** [comment]
oli has your friend reviewed the issue?

**Comment by dromer** [comment]
@RustoMCSpit can you please be less pushy to developers and let them work in their own schedule?

I'm sure we will get an update when there is an update. No need to nag about it.

**Comment by RustoMCSpit** [comment]
> @RustoMCSpit can you please be less pushy to developers and let them work in their own schedule?
> 
> I'm sure we will get an update when there is an update. No need to nag about it.

sorry

**Comment by olilarkin** [comment]
Thanks @dromer. @RustoMCSpit I will have to block you if you behave like this, and other open source maintainers will do the same.  It's not OK. If you want Linux support so badly for iPlug2 you are free to do that work yourself.

**Comment by CodesoundR** [comment]
Hi,

I have read about some misunderstandings about the Linux port of iplug2 but I would not like to open a new thread on this topic. So I apologize in advance I just wanted to warn that iPlug2 is used ALSO for the development of the beautiful Instruments of FullBucketMusic that I personally contacted to ask if he ported these in the Linux world; he kindly replied informing me that he can't do it until iPlug2 has Linux support.  I really hope that soon there will be this support. I am not able to do anything about it, I can only ask you who are the experts..... These instruments are wonderful and I think you know them too..... I thank you in advance if only for the patience shown in reading this prayer of mine..... Thanks again for everything....

regards

[](https://www.fullbucket.de/music/vst.html)

**Comment by tkna91** [comment]
@CodesoundR I am also following this isssue for the same reason as you.
My one hope is that the fullbucketmusic plugin will work on Linux without yabridge. I am really eager for it.

**Comment by CodesoundR** [comment]
> @CodesoundR I am also following this isssue for the same reason as you. My one hope is that the fullbucketmusic plugin will work on Linux without yabridge. I am really eager for it.

Hi, 
From what I seem to have understood, there are many projects related to this library. I'm not a programmer, but only lately I've understood the importance of this kind of libraries (that fortunately there are!!!). I use some native Linux tools, but it's not all there. I manage to have what is missing thanks to Wine-HQ staging, but entrusting various licenses and the stability of my system to a project that often makes it completely unstable, and hoping that with each update everything works as before (which never happened), I can't. In fact, I often find myself asking the developer directly if he develops his software for Linux (I prefer the.clap extension, but the VST3 format is also fine). I've been doing this for some time: I've received very positive as well as negative responses. And so I'm tracing a new path using tools that gradually replace those of windows and that are native to Linux. thanks to the email sent to FullBucketMusic, I understood the existence of this library, so I asked. I would be really extremely HAPPY to see the port of this one to linux..... I think it will open a huge door of software development for Linux as well. I hope in the goodness and reasonableness of the developers. Thank you very much for all your work.....
Regards

**Comment by RustoMCSpit** [comment]
> How about we break down the architecture of the Linux support feature into smaller tasks/stages that the community could possibly help out with?

I am not sure of what has been done already, but this would be a general idea:

- [ ] Build System
  - [ ] CMake
  - [ ] Compiler: GCC or Clang
  - [ ] Set up dependencies
    - [ ] build-essential for basic compilation tools
    - [ ] libjack-dev and libjack-jackd2-dev for audio I/O
    - [ ] libgtk-3-dev for GUI
    - [ ] libx11-xcb-dev for X11 support
- [ ] IGraphics
  - [ ] Backend: NanoVG or Skia
    - [ ] drawing operations
    - [ ] text rendering
    - [ ] event handling
    - [ ] Add text entry support
    - [ ] Add menu support
- [ ] Linux windowing systems
  - [ ] X11 client
  - [ ] Wayland client
- [ ] plug-in APIs on Linux
  - [ ] SDK and API Support for LV2
  - [ ] SDK and API Support for VST3
  - [ ] SDK and API Support for CLAP
  - [ ] SDK and API Support for WAM
- [ ] Set up audio and MIDI I/O for standalone apps
  - [ ] ALSA
  - [ ] JACK
  - [ ] PulseAudio
  - [ ] Implement MIDI I/O
- [ ] Adapt platform-specific code for Linux
  - [ ] Use POSIX-compliant APIs for file handling and threading
  - [ ] Add inotify-based file watching for enhanced functionality

**Comment by RustoMCSpit** [comment]
@olilarkin anything else need adding?

**Comment by RustoMCSpit** [comment]
I didn't know what distribution methods iPlug was going to under take so I just threw in a lot of them for @olilarkin to decide on!

**Comment by RustoMCSpit** [comment]
> Hello people! Since [@olilarkin](https://github.com/olilarkin) has already "blessed" this by opening [Linux support](https://github.com/orgs/iPlug2/projects/11), and since we already have some good PoCs early on, how about we break down the architecture of the Linux support feature into smaller tasks/stages that the community could possibly help out with? Lots of people seem to be doing great things with this library, and I'm sure we could scrape together the dev resources to bring it to Linux native - as long as there's some outline of the work that suits the project maintainers' vision.

we need to make some issue requests for everyone and decide on a tech stack. the check list i provided was more of a general idea than specifics. do you or bindingnews have any modifications you want to make to the check list or should we start making issue requests and tagging them to the linux support project https://github.com/orgs/iPlug2/projects/11

**Comment by olilarkin** [comment]
RustoMCSpit can you give me some context? IIRC you aren't a dev yourself, but you seem to be trying to get this moving. I am extremely busy at the moment but I hope to be able to think about iPlug2 and address the CMake and Linux support at some point, checkout @Bindernews latest work etc. I would also still be very happy to find the right person/people to hire to do this work, if anyone knows the right person who is qualified and interested/has time.

I would rather that you don't start filing issues if you are not a dev doing the work.

**Comment by sstillwell** [comment]
I have proof-of-concept builds of Event Horizon running on Linux using Skia and VST3 format.  I have enough of IGraphics running to provide a typical plugin UI, but I don't have keyboard input yet.  For mouse navigation it's pretty much feature parity with Windows or macOS.

Unfortunately my repos don't use iPlug2 out of the box, and my build scripts using CMake and Python diverge pretty sharply from what you have (we developed it before this project got started in earnest on CMake).  Not sure how I could package it for you to accept even if you wanted it...I'm not unwilling, however.

I build on an older Debian build for maximal glibc compatibility with end users.

I haven't really thought about building LV2, but CLAP is on  my roadmap (no rhyming puns intended).

**Comment by RangerRick** [comment]
I have gotten everything running with Linux for our plugin using X11, with both VST3 (tested) and CLAP (not tested). I did all the development in one big blob but was able to extract out the purely-linux stuff and have a working prototype. I went with X11 just because it's most likely to work everywhere, and hooo boy, was that a mistake. Plugin resizing code is a mess. 😅 

I'll make a PR with what I've got and you can tell me what you think.

**Comment by JsamsIvyTech** [comment]
Can I help in some way
